app/src/main/java/com/github/geohunt/app/model/database/firebase/FirebaseDatabase.kt

```diff
@@ -140,6 +142,62 @@ class FirebaseDatabase(activity: Activity) : Database {
     override fun getNearbyChallenge(location: Location): Task<List<Challenge>> {
         TODO()
     }
+
+    override fun getFollowersOf(uid: String): Task<Map<String, Boolean>> {
+        return dbFollowersRef.child(uid)
+            .get()
+            .thenMap { snapshot ->
+                snapshot.toMap<Boolean>().withDefault { false }
+            }
+    }
+
+    /**
+     * Updates the database to (un)follow the given users.
+     *
+     * @note To ensure easier data querying, the data must be synchronized at 3 places in the database :
+     *       - In the follower's follow list, the followee should be added/removed,
+     *       - The followee's counter should be incremented/decremented,
+     *       - The follower should be added/removed to the followee's follower list.
+     */
+    private suspend fun doFollow(follower: String, followee: String, follow: Boolean = true) {
+        val followerListRef = dbUserRef.child(follower).child("followList")
+        val counterRef = dbUserRef.child(followee).child("followers")
+        val follows = dbFollowersRef.child(followee)
+
+        val defaultMap = emptyMap<String, Boolean>().withDefault { false }
+        val followerList = followerListRef.queryAs<Map<String, Boolean>>() ?: defaultMap
+        val followedCounter = counterRef.queryAs<Int>() ?: 0
+        val followsPairs = follows.queryAs<Map<String, Boolean>>() ?: defaultMap
+
+        // Abort if the user already follows the followee
+        // or if the user tries to unfollow someone not followed
+        if ((follow && followerList[followee] == true)
+            || (!follow && followerList[followee] != true)) {
+            return
+        }
+
+        Tasks.whenAll(
+            // Update the follower's follow list
+            followerListRef.setValue(if (follow) (followerList + (followee to true)) else followerList - followee),
+
+            // Update the follow counter of the followee
+            counterRef.setValue(if (follow) followedCounter + 1 else max(followedCounter - 1, 0)),
+
+            // Update the follows relationship
+            follows.setValue(if (follow) followsPairs + (follower to true) else followsPairs - follower)
+        ).await()
+    }
```

Comment:
This function is not concurrent at all and result is highly undefined behavior. I suggest using [transaction](https://firebase.google.com/docs/database/android/read-and-write#save_data_as_transactions) to perform such operation on Firebase. It should take ~30 minutes or so and I can help you implement it (actually I am curious myself)

Reply:
That's true. While the current implementation uses the Tasks.whenAll() function to wait for all tasks to complete, this approach may not scale well if the number of tasks increases significantly.

Using a transaction to perform this operation on Firebase is a good solution to ensure data consistency and prevent race conditions. Transactions provide an atomic way to read and write data, allowing you to ensure that your database updates are always consistent and correct.

Reply:
I was originally looking to batch the write operations, and only found out the [Firestore version of transactions](https://firebase.google.com/docs/firestore/manage-data/transactions?hl=en)...

Didn't know that they were also implemented for the Realtime Database. I'll implement the transaction then 😅

Reply:
I've opened issue #88 to implement transactions, as it would require some heavy work that I think is not the priority.

Reply:
Allright